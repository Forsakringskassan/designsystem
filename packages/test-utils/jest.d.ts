export * from "./dist/types/jest";

declare global {
    namespace jest {
        interface Matchers<R> {
            /**
             * Expects element to have focus.
             */
            toHaveFocus(): R;
        }
    }
}
