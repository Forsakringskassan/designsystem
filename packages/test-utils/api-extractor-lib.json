{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
