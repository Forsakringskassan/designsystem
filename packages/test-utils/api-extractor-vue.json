{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
