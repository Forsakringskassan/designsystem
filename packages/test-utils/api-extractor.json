{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
