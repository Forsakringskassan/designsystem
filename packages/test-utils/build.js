const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
