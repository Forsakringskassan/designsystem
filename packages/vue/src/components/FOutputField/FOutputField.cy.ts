import { defineComponent } from "vue";
import {
    SizeWrapper,
    sizeWrapperWidth,
    sizeWrapperHeight,
    DensityWrapper,
    densityWrapperWidth,
    densityWrapperHeight,
} from "@fkui/test-utils/vue";
import { FOutputFieldPageobject } from "../../cypress";
import { FTextField } from "../FTextField";
import FOutputField from "./FOutputField.vue";

const TestComponent = defineComponent({
    template: /* HTML */ `
        <size-wrapper>
            <template #default="{ variant }">
                <div class="row">
                    <div class="col col--md-6">
                        <f-text-field
                            v-model="value"
                            :id="'awesome-id--' + variant"
                        >
                            Värde
                        </f-text-field>
                    </div>
                    <div class="col col--md-6" v-test="'output-field'">
                        <f-output-field :for="'awesome-id--' + variant">
                            <template #label> Beräknat värde </template>
                            <template #default> {{ value + 1 }} </template>
                        </f-output-field>
                    </div>
                </div>
            </template>
        </size-wrapper>
    `,
    components: {
        SizeWrapper,
        FOutputField,
        FTextField,
    },
    data() {
        return {
            value: 3,
        };
    },
});
it("should provide a page object that can access any necessary elements", () => {
    cy.mount(TestComponent);
    const output = new FOutputFieldPageobject("[data-test] .output-field");
    output.label.el().eq(0).should("have.trimmedText", "Beräknat värde");
    output.body().eq(0).should("have.trimmedText", "4");
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
it.skip(`should match screenshot`, () => {
    cy.viewport(sizeWrapperWidth, sizeWrapperHeight);
    cy.mount(TestComponent);
    cy.toMatchScreenshot();
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
describe.skip("density", () => {
    const DensityComponent = defineComponent({
        template: /* HTML */ `
            <density-wrapper>
                <f-text-field v-model="value"> Värde </f-text-field>
                <f-output-field>
                    <template #label> Beräknat värde </template>
                    <template #default> {{value + 1}} </template>
                </f-output-field>
            </density-wrapper>
        `,
        components: {
            DensityWrapper,
            FTextField,
            FOutputField,
        },
        data() {
            return { value: 1 };
        },
    });

    it(`should be densified`, () => {
        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.toMatchScreenshot();
    });
});
