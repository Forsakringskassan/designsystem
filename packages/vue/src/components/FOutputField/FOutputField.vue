<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FLabel } from "../FLabel";

export default defineComponent({
    name: "FOutputField",
    components: {
        FLabel,
    },
    inheritAttrs: false,
    props: {
        /**
         * Associate the output field with one or more id's of the elements
         * contributing or affecting the result of this field. Multiple id's can
         * be set with a space-separated string or an array of strings.
         */
        for: {
            type: [String, Array],
            required: true,
        },
        /**
         * The id for the output id attribute.
         * The id for the label for attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
    },
    computed: {
        htmlFor(): string {
            return Array.isArray(this.for) ? this.for.join(" ") : this.for;
        },
    },
});
</script>

<template>
    <div class="output-field">
        <f-label :for="id">
            <template #default>
                <!-- @slot Slot for label content. -->
                <slot name="label"></slot>
            </template>
            <!-- @slot Slot for tooltip. -->
            <template v-if="$slots.tooltip" #tooltip>
                <!-- @slot Slot primarily for tooltip. -->
                <slot name="tooltip"></slot>
            </template>
        </f-label>
        <output :id="id" :for="htmlFor" class="output-field__output" v-bind="$attrs">
            <!-- @slot Slot for output field content. -->
            <slot name="default"></slot>
        </output>
    </div>
</template>
