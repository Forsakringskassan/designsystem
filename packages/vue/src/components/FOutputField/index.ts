export { default as FOutputField } from "./FOutputField.vue";
