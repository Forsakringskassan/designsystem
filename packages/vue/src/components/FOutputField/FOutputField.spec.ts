import { VueWrapper, mount } from "@vue/test-utils";
import FOutputField from "./FOutputField.vue";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FOutputField, {
        attrs: { ...attrs },
        props: { for: "", ...props },
        slots: { default: "1234", ...slots },
    });
}

describe("snapshots", () => {
    it("should match snapshot with output and for attribute", () => {
        const wrapper = createWrapper({
            props: {
                for: "inputid",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with output and two for attributes", () => {
        const wrapper = createWrapper({
            props: {
                for: "inputid inputid2",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with label, tooltip, output and for attribute", () => {
        const wrapper = createWrapper({
            slots: { label: "Summa", tooltip: "TOOLTIP" },
            props: {
                for: "inputid",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("props", () => {
    describe("id", () => {
        it("should set matching id on <label> and <output>", () => {
            expect.assertions(2);
            const id = "foobar";
            const wrapper = mount(FOutputField, {
                props: {
                    for: "my-awesome-id",
                    id,
                },
            });
            const label = wrapper.get("label");
            const output = wrapper.get("output");
            expect(label.attributes("for")).toEqual(id);
            expect(output.attributes("id")).toEqual(id);
        });

        it("should autogenerate id if not explicitly given", () => {
            expect.assertions(1);
            const wrapper = mount(FOutputField, {
                props: {
                    for: "my-awesome-id",
                },
            });
            const label = wrapper.get("label");
            const output = wrapper.get("output");
            const id = output.attributes("id");
            expect(label.attributes("for")).toEqual(id);
        });
    });

    describe("for", () => {
        it("should handle space-separated string", () => {
            expect.assertions(1);
            const id = "foobar";
            const wrapper = mount(FOutputField, {
                props: {
                    for: id,
                },
            });
            const output = wrapper.get("output").element as HTMLOutputElement;
            expect(Array.from(output.htmlFor)).toEqual([id]);
        });

        it("should handle array", () => {
            expect.assertions(1);
            const id1 = "foo";
            const id2 = "bar";
            const wrapper = mount(FOutputField, {
                props: {
                    for: [id1, id2],
                },
            });
            const output = wrapper.get("output").element as HTMLOutputElement;
            expect(Array.from(output.htmlFor)).toEqual([id1, id2]);
        });

        it("should handle mixing array and space-separated string", () => {
            expect.assertions(1);
            const id1 = "foo bar";
            const id2 = "baz";
            const wrapper = mount(FOutputField, {
                props: {
                    for: [id1, id2],
                },
            });
            const output = wrapper.get("output").element as HTMLOutputElement;
            expect(Array.from(output.htmlFor)).toEqual(["foo", "bar", "baz"]);
        });
    });
});

describe("attributes", () => {
    it("should pass attributes", () => {
        const wrapper = createWrapper({
            attrs: {
                for: "my-awesome-id",
                foo: "bar",
            },
        });
        const output = wrapper.get("output");
        expect(output.attributes("foo")).toBe("bar");
    });
});
