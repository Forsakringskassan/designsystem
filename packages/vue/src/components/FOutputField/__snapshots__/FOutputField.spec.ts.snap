// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with label, tooltip, output and for attribute 1`] = `
<div
  class="output-field"
>
  <div>
    <div
      class="tooltip-before"
    >
      <label
        class="label tooltip-before__label"
        for="fkui-vue-element-0001"
      >
        
        
        Summa
        
        
      </label>
    </div>
    
    
    TOOLTIP
    
    
    <!--v-if-->
  </div>
  <output
    class="output-field__output"
    for="inputid"
    id="fkui-vue-element-0001"
  >
    
    1234
    
  </output>
</div>
`;

exports[`snapshots should match snapshot with output and for attribute 1`] = `
<div
  class="output-field"
>
  <label
    class="label"
    for="fkui-vue-element-0001"
  >
    
    
    
    
    <!--v-if-->
  </label>
  <output
    class="output-field__output"
    for="inputid"
    id="fkui-vue-element-0001"
  >
    
    1234
    
  </output>
</div>
`;

exports[`snapshots should match snapshot with output and two for attributes 1`] = `
<div
  class="output-field"
>
  <label
    class="label"
    for="fkui-vue-element-0001"
  >
    
    
    
    
    <!--v-if-->
  </label>
  <output
    class="output-field__output"
    for="inputid inputid2"
    id="fkui-vue-element-0001"
  >
    
    1234
    
  </output>
</div>
`;
